### ISSUE TITLE
autocrafter: if there isn't enough room in the output inventory, replacements may be lost

### ISSUE THREAD

**Issue by fluxionary** [mt-mods/pipeworks#61]
expected behavior: crafting should not occur unless there is room for the output and all replacements simultaneously. 

hopefully these images are enough to demonstrate the problem:

![image](https://user-images.githubusercontent.com/25628292/216832639-e2af43d4-f966-4ffd-8974-b59c36013add.png)

![image](https://user-images.githubusercontent.com/25628292/216832664-a937eed8-65f7-483a-bd93-7a29ed807346.png)

**Comment by S-S-X** [comment]
Wasn't this one already fixed once? Not sure, might also be just in game workarounds... or talk about fixing it...

**Comment by SwissalpS** [comment]
core provides ``inv:room_for_item(list, stack)`` but not ``inv:room_for_items(list, {stack1, ... stackN })`` so we either put our energy in adding that to core ...  

... or meanwhile I have a sketch that could work until then.

There is this situation that results a bit messy, but at least no buckets go missing:
![Captura de pantalla de 2023-06-11 05-11-56](https://github.com/mt-mods/pipeworks/assets/161979/fd94ecfb-6a73-4d19-99e3-5e40669d8209)
![Captura de pantalla de 2023-06-11 05-12-16](https://github.com/mt-mods/pipeworks/assets/161979/31449ec9-ca90-46ee-92c4-c4a9e1780191)

Basically I'm proposing to keep an undo history of items put in dst and then taking them out again if something didn't fit.

**Comment by SwissalpS** [comment]
coconut issue not solved with this. Axe is not in the ``decrement_input`` part of the craft table.

**Comment by SwissalpS** [comment]
A cleaner approach would be to clone dst inv, apply outputs in that sandbox and on success clone back and let garbage collector clear out the clone.  
I fear that would add quite some overhead.

I'm trying to avoid the code-intensive variant of indexing dst and trying to figure out if all output materials fit.

**Comment by SwissalpS** [comment]
#70 uses sandbox inventory. UX and code is cleaner too.

since MT is single threaded, I guess it is safe to use the same sandbox for all autocrafters.

**Comment by SwissalpS** [comment]
#71 uses a slimmed down fake inventory which indexes destination to determine if output items all fit.
